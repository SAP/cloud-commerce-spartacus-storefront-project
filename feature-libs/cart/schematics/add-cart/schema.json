{
  "$schema": "http://json-schema.org/schema",
  "id": "CartSchematics",
  "title": "Cart Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "lazy": {
      "type": "boolean",
      "description": "Lazy load the Saved cart feature.",
      "default": true
    },
    "features": {
      "type": "array",
      "uniqueItems": true,
      "items": {
        "enum": ["Saved-cart"],
        "type": "string"
      },
      "default": ["Saved-cart"],
      "x-prompt": "Which features would you like to setup?"
    }
  },
  "required": []
}
