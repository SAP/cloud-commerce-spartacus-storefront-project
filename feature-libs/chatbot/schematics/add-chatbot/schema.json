{
  "$schema": "http://json-schema.org/schema",
  "$id": "ChatBotSchematics",
  "title": "ChatBot Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "lazy": {
      "type": "boolean",
      "description": "Lazy load the chat bot features.",
      "default": true
    },
    "features": {
      "type": "array",
      "uniqueItems": true,
      "default": ["ChatBot"]
    }
  },
  "required": []
}
