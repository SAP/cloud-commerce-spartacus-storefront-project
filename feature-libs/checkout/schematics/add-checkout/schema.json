{
  "$schema": "http://json-schema.org/schema",
  "$id": "CheckoutSchematics",
  "title": "Checkout Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "lazy": {
      "type": "boolean",
      "description": "Lazy load the Checkout feature.",
      "default": true
    },
    "features": {
      "type": "array",
      "uniqueItems": true,
      "items": {
        "enum": [
          "Checkout",
          "Checkout-B2B",
          "Checkout-Scheduled-Replenishment"
        ],
        "type": "string"
      },
      "default": ["Checkout"],
      "x-prompt": {
        "message": "Which features would you like to set up from the Checkout library?",
        "type": "list",
        "items": [
          {
            "value": "Checkout",
            "label": "Checkout"
          },
          {
            "value": "Checkout-B2B",
            "label": "Checkout (b2b features)"
          },
          {
            "value": "Checkout-Scheduled-Replenishment",
            "label": "Checkout Scheduled Replenishment (b2b feature)"
          }
        ]
      }
    }
  },
  "required": []
}
