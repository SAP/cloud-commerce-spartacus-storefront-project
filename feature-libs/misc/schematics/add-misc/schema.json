{
  "$schema": "http://json-schema.org/schema",
  "id": "MiscSchematics",
  "title": "Misc Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "lazy": {
      "type": "boolean",
      "description": "Lazy load the misc features.",
      "default": true
    },
    "features": {
      "type": "array",
      "uniqueItems": true,
      "items": {
        "enum": ["Storefinder"],
        "type": "string"
      },
      "default": ["Storefinder"],
      "x-prompt": "Which features would you like to setup?"
    }
  },
  "required": []
}
