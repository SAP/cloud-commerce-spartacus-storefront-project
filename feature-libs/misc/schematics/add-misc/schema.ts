export interface Schema {
  project: string;
  lazy: boolean;
  features: string[];
}
