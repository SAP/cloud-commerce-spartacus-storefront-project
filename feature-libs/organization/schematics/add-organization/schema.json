{
  "$schema": "http://json-schema.org/schema",
  "id": "OrganizationSchematics",
  "title": "Organization Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    }
  },
  "required": []
}
