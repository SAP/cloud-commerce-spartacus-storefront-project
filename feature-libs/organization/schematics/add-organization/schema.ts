export interface Schema {
  project: string;
  lazy: boolean;
}
