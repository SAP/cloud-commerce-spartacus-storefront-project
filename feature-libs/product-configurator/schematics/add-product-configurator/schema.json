{
  "$schema": "http://json-schema.org/schema",
  "id": "ProductConfiguratorSchematics",
  "title": "Product Configurator Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "lazy": {
      "type": "boolean",
      "description": "Lazy load the product configurator features.",
      "default": true
    },
    "features": {
      "type": "array",
      "uniqueItems": true,
      "items": {
        "enum": ["CPQ configurator", "Textfield configurator"],
        "type": "string"
      },
      "x-prompt": "Which product configurator features besides variant configurator would you like to setup?"
    }
  },
  "required": []
}
