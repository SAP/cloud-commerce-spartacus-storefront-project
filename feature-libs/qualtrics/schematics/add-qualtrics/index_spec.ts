/// <reference types="jest" />

import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import {
  Schema as ApplicationOptions,
  Style,
} from '@schematics/angular/application/schema';
import { Schema as WorkspaceOptions } from '@schematics/angular/workspace/schema';
import {
  CLI_QUALTRICS_FEATURE,
  LibraryOptions as SpartacusQualtricsOptions,
  SpartacusOptions,
} from '@spartacus/schematics';
import * as path from 'path';

const collectionPath = path.join(__dirname, '../collection.json');
const qualtricsModulePath =
  'src/app/spartacus/features/qualtrics/qualtrics-feature.module.ts';
const scssFilePath = 'src/styles/spartacus/qualtrics-embedded-feedback.scss';

// TODO: Improve tests after lib-util test update
describe('Spartacus Qualtrics schematics: ng-add', () => {
  const schematicRunner = new SchematicTestRunner('schematics', collectionPath);

  let appTree: UnitTestTree;

  const workspaceOptions: WorkspaceOptions = {
    name: 'workspace',
    version: '0.5.0',
  };

  const appOptions: ApplicationOptions = {
    name: 'schematics-test',
    inlineStyle: false,
    inlineTemplate: false,
    routing: false,
    style: Style.Scss,
    skipTests: false,
    projectRoot: '',
  };

  const spartacusDefaultOptions: SpartacusOptions = {
    project: 'schematics-test',
    configuration: 'b2c',
    lazy: true,
    features: [],
  };

  const defaultFeatureOptions: SpartacusQualtricsOptions = {
    project: 'schematics-test',
    lazy: true,
    features: [CLI_QUALTRICS_FEATURE],
  };

  beforeEach(async () => {
    schematicRunner.registerCollection(
      '@spartacus/schematics',
      '../../projects/schematics/src/collection.json'
    );

    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'workspace',
        workspaceOptions
      )
      .toPromise();
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'application',
        appOptions,
        appTree
      )
      .toPromise();
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@spartacus/schematics',
        'ng-add',
        { ...spartacusDefaultOptions, name: 'schematics-test' },
        appTree
      )
      .toPromise();
  });

  describe('When no features are provided', () => {
    beforeEach(async () => {
      appTree = await schematicRunner
        .runSchematicAsync(
          'ng-add',
          { ...defaultFeatureOptions, features: [] },
          appTree
        )
        .toPromise();
    });

    it('should not create the feature module', () => {
      const featureModule = appTree.readContent(qualtricsModulePath);
      expect(featureModule).toBeFalsy();
    });
    it('should not add the feature to the feature module', () => {
      const spartacusFeaturesModulePath = appTree.readContent(
        'src/app/spartacus/spartacus-features.module.ts'
      );
      expect(spartacusFeaturesModulePath).toMatchSnapshot();
    });
  });

  describe('Qualtrics feature', () => {
    describe('styling', () => {
      beforeEach(async () => {
        appTree = await schematicRunner
          .runSchematicAsync('ng-add', defaultFeatureOptions, appTree)
          .toPromise();
      });

      it('should create a proper scss file', () => {
        const scssContent = appTree.readContent(scssFilePath);
        expect(scssContent).toMatchSnapshot();
      });

      it('should update angular.json', async () => {
        const content = appTree.readContent('/angular.json');
        expect(content).toMatchSnapshot();
      });
    });

    describe('eager loading', () => {
      beforeEach(async () => {
        appTree = await schematicRunner
          .runSchematicAsync(
            'ng-add',
            { ...defaultFeatureOptions, lazy: false },
            appTree
          )
          .toPromise();
      });

      it('should import appropriate modules', async () => {
        const qualtricsModule = appTree.readContent(qualtricsModulePath);
        expect(qualtricsModule).toContain(
          `import { QualtricsRootModule } from "@spartacus/qualtrics/root";`
        );
        expect(qualtricsModule).toContain(
          `import { QualtricsModule } from "@spartacus/qualtrics";`
        );
      });

      it('should not contain lazy loading syntax', async () => {
        const qualtricsModule = appTree.readContent(qualtricsModulePath);
        expect(qualtricsModule).not.toContain(
          `import('@spartacus/qualtrics').then(`
        );
      });
    });

    describe('lazy loading', () => {
      beforeEach(async () => {
        appTree = await schematicRunner
          .runSchematicAsync('ng-add', defaultFeatureOptions, appTree)
          .toPromise();
      });

      it('should import QualtricsRootModule and contain the lazy loading syntax', async () => {
        const qualtricsModule = appTree.readContent(qualtricsModulePath);
        expect(qualtricsModule).toContain(
          `import { QualtricsRootModule, QUALTRICS_FEATURE } from "@spartacus/qualtrics/root";`
        );
        expect(qualtricsModule).toContain(
          `import('@spartacus/qualtrics').then(`
        );
      });

      it('should not contain the QualtricsModule import', () => {
        const qualtricsModule = appTree.readContent(qualtricsModulePath);
        expect(qualtricsModule).not.toContain(
          `import { QualtricsModule } from "@spartacus/qualtrics";`
        );
      });
    });
  });
});
