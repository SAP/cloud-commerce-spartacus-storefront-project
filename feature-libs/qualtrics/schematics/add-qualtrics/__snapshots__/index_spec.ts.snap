// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spartacus Qualtrics schematics: ng-add Qualtrics feature eager loading should import appropriate modules 1`] = `
"import { NgModule } from '@angular/core';
import { QualtricsModule } from \\"@spartacus/qualtrics\\";
import { QualtricsRootModule } from \\"@spartacus/qualtrics/root\\";

@NgModule({
  declarations: [],
  imports: [
    QualtricsRootModule,
    QualtricsModule
  ]
})
export class QualtricsFeatureModule { }
"
`;

exports[`Spartacus Qualtrics schematics: ng-add Qualtrics feature general setup should add the feature using the lazy loading syntax 1`] = `
"import { NgModule } from '@angular/core';
import { CmsConfig, provideConfig } from \\"@spartacus/core\\";
import { QualtricsRootModule, QUALTRICS_FEATURE } from \\"@spartacus/qualtrics/root\\";

@NgModule({
  declarations: [],
  imports: [
    QualtricsRootModule
  ],
  providers: [provideConfig(<CmsConfig>{
    featureModules: {
      [QUALTRICS_FEATURE]: {
        module: () =>
          import('@spartacus/qualtrics').then((m) => m.QualtricsModule),
      },
    }
  })]
})
export class QualtricsFeatureModule { }
"
`;

exports[`Spartacus Qualtrics schematics: ng-add Qualtrics feature general setup styling should create a proper scss file 1`] = `"@import \\"@spartacus/qualtrics\\";"`;

exports[`Spartacus Qualtrics schematics: ng-add Qualtrics feature general setup styling should update angular.json 1`] = `
"{
  \\"$schema\\": \\"./node_modules/@angular/cli/lib/config/schema.json\\",
  \\"version\\": 1,
  \\"newProjectRoot\\": \\"\\",
  \\"projects\\": {
    \\"schematics-test\\": {
      \\"projectType\\": \\"application\\",
      \\"schematics\\": {
        \\"@schematics/angular:component\\": {
          \\"style\\": \\"scss\\"
        }
      },
      \\"root\\": \\"\\",
      \\"sourceRoot\\": \\"src\\",
      \\"prefix\\": \\"app\\",
      \\"architect\\": {
        \\"build\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:browser\\",
          \\"options\\": {
            \\"outputPath\\": \\"dist/schematics-test\\",
            \\"index\\": \\"src/index.html\\",
            \\"main\\": \\"src/main.ts\\",
            \\"polyfills\\": \\"src/polyfills.ts\\",
            \\"tsConfig\\": \\"tsconfig.app.json\\",
            \\"aot\\": true,
            \\"assets\\": [
              \\"src/favicon.ico\\",
              \\"src/assets\\"
            ],
            \\"styles\\": [
              \\"src/styles.scss\\",
              \\"src/styles/spartacus/qualtrics-embedded-feedback.scss\\"
            ],
            \\"scripts\\": []
          },
          \\"configurations\\": {
            \\"production\\": {
              \\"fileReplacements\\": [
                {
                  \\"replace\\": \\"src/environments/environment.ts\\",
                  \\"with\\": \\"src/environments/environment.prod.ts\\"
                }
              ],
              \\"optimization\\": true,
              \\"outputHashing\\": \\"all\\",
              \\"sourceMap\\": false,
              \\"extractCss\\": true,
              \\"namedChunks\\": false,
              \\"extractLicenses\\": true,
              \\"vendorChunk\\": false,
              \\"buildOptimizer\\": true,
              \\"budgets\\": [
                {
                  \\"type\\": \\"initial\\",
                  \\"maximumWarning\\": \\"2mb\\",
                  \\"maximumError\\": \\"5mb\\"
                },
                {
                  \\"type\\": \\"anyComponentStyle\\",
                  \\"maximumWarning\\": \\"6kb\\",
                  \\"maximumError\\": \\"10kb\\"
                }
              ]
            }
          }
        },
        \\"serve\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:dev-server\\",
          \\"options\\": {
            \\"browserTarget\\": \\"schematics-test:build\\"
          },
          \\"configurations\\": {
            \\"production\\": {
              \\"browserTarget\\": \\"schematics-test:build:production\\"
            }
          }
        },
        \\"extract-i18n\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:extract-i18n\\",
          \\"options\\": {
            \\"browserTarget\\": \\"schematics-test:build\\"
          }
        },
        \\"test\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:karma\\",
          \\"options\\": {
            \\"main\\": \\"src/test.ts\\",
            \\"polyfills\\": \\"src/polyfills.ts\\",
            \\"tsConfig\\": \\"tsconfig.spec.json\\",
            \\"karmaConfig\\": \\"karma.conf.js\\",
            \\"assets\\": [
              \\"src/favicon.ico\\",
              \\"src/assets\\"
            ],
            \\"styles\\": [
              \\"src/styles.scss\\",
              \\"src/styles/spartacus/qualtrics-embedded-feedback.scss\\"
            ],
            \\"scripts\\": []
          }
        },
        \\"lint\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:tslint\\",
          \\"options\\": {
            \\"tsConfig\\": [
              \\"tsconfig.app.json\\",
              \\"tsconfig.spec.json\\",
              \\"e2e/tsconfig.json\\"
            ],
            \\"exclude\\": [
              \\"**/node_modules/**\\"
            ]
          }
        },
        \\"e2e\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:protractor\\",
          \\"options\\": {
            \\"protractorConfig\\": \\"e2e/protractor.conf.js\\",
            \\"devServerTarget\\": \\"schematics-test:serve\\"
          },
          \\"configurations\\": {
            \\"production\\": {
              \\"devServerTarget\\": \\"schematics-test:serve:production\\"
            }
          }
        }
      }
    }
  },
  \\"defaultProject\\": \\"schematics-test\\"
}"
`;
